net: l2: ieee802154: unified timestamp preparation

fgrandel opened the pull request:
This pull request works towards ongoing realization of the following RFCs:

* #50336
* Layer 1 (clock sources) and layer 2 (syntonization) of a [high-level clock subsystem](https://github.com/zephyrproject-rtos/zephyr/issues/19030#issuecomment-1597226731)
* [Improved cross-vendor support of the IEEE 802.15.4 driver and L2 stack](https://github.com/zephyrproject-rtos/zephyr/pull/58438#discussion_r1220059230)

~~The kernel part of this PR needs to be approved in a separate PR (#60400) before this PR can be merged. It is provided as an example application of ktime_t and for preliminary preview wrt the above cited RFCs.~~

UPDATE: It has been proposed by myself and @cfriedt to decouple the two PRs as the kernel part needs more work and shouldn't block the unification of timestamps in the network subsystem - which is useful on its own.

This PR is a precondition for fixing #59245.

Commit:
doc: drivers: ieee802154: radio API

Improves the documentation of the IEEE 802.15.4 radio API.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: consistent high res timestamps

The IEEE 802.15.4 API and networking subsystem were using several
inconsistent timestamp resolutions and types. This change defines all
timestamps with nanosecond resolution and reduces the number of
available types to represent timestamps to two:
* `struct net_ptp_time` for PTP timestamps
* `net_time_t` for all other high resolution timestamps

All timestamps (including PTP timestamps) are now referred to a
"virtual" local network subsystem clock source based on the well-defined
types above. It is the responsibility of network subsystem L2/driver
implementations (notably Ethernet and IEEE 802.15.4 L2 stacks) to ensure
consistency of all timestamps and radio timer values exposed by the
driver API to such a network subsystem uptime reference clock
independent of internal implementation details.

The "virtual" network clock source may be implemented based on arbitrary
hardware peripherals (e.g. a coarse low power RTC counter during sleep
time plus a high resolution/high precision radio timer while receiving
or sending). Such implementation details must be hidden from API
clients, as if the driver used a single high resolution clock source
instead.

For IEEE 802.15.4, whenever timestamps refer to packet send or receive
times, they are measured when the end of the IEEE 802.15.4 SFD (message
timestamp point) is present at the local antenna (reference plane).

Due to its limited range of ~290 years, net_time_t timestamps (and
therefore net_pkt timestamps and times) must not be used to represent
absolute points in time referred to an external epoch independent of
system uptime (e.g.  UTC, TAI, PTP, NTP, ...).

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
doc: net: ptp: PTP structs

Improves documentation of PTP structs and explains basic underlying
concepts to increase the probability that these structs will be used
correctly and consistently.

Also introduces references to the underlying specifications.

Note: We currently (ab)use the PTP structs for timestamps in the IEEE
802.15.4 context for which they are undefined. It is also not ideal that
the generic `struct net_pkt` depends directly on PTP. Future changes
will therefore have to remove the reference to PTP structs in net_pkt
and replace them by net_time_t. Clients will then have to convert these
to PTP structures if required.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
tests: subsys: openthread: support TXTIME

Adds a test suite configuration that enables TXTIME.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
tests: subsys: openthread: simplify configure mocks

Removes redundant fakes.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
tests: subsys: openthread: support CSL

Introduces coverage for OpenThread CSL platform API as far as channel
samples are concerned.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: nRF5: fix return type

Adapts a return type to the API specification. The changed return type
is not referenced anywhere so it can be changed without breaking
backwards compatibility.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: introduce scalar nanosecond time representation

Introduces a well-defined intermediate concept of syntonized scalar
nanosecond resolution time with overflow protection above low-level
counters/cycles/ticks and below higher level time abstractions
(timescales, calenders, etc.).

The rationale of this type has been extensively documented and
contrasted to already existing time representations to ensure that it
fills a well defined gap without overlap.

This change prepares for later changes in this change set that will
unify the usage of time across the network subsystem (RX/TX timestamps,
timed TX, CSL, scheduled reception windows, (g)PTP integration, etc.).

The type is EXPERIMENTAL and named net_time_t while it is not used in a
larger clock subsystems, the details of which are still being discussed
(see #60400 for details).

See
https://github.com/zephyrproject-rtos/zephyr/issues/19030#issuecomment-1597226731
for its embedding in a larger clock subsystem architecture relevant to
the network stack, IEEE 802.15.4 and the POSIX roadmap.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

fgrandel commented:
This PR can now be reviewed. It has been proposed in #60400 to not couple the development of the larger clock subsystem to this change as this change is useful in itself and shouldn't be held back due to other PRs depending on it.

@nandojve @jukkar Probably also interesting to you as this introduces a unified concept for all timestamps and time spans in the network stack.

fgrandel commented:
@edmont Heads up to you as this is the change set that prepares for the fix of #59245 which we're currently working on together.

fgrandel commented:
@edmont @rlubos @jukkar I'm not very happy with the ktime_t name - it was named like in Linux as it is the same concept and was meant to be part of a larger clock subsystem but that will take time. I like the idea to have a common well-specified type for scalar nanosecond timestamps in the network subsystem, though, and I believe that this PR shows that the idea works rather nicely. Maybe you have a better idea how to name this type? It should still be possible to rename it later when it is replaced by something more generic.

net_timestamp_t? net_ns_uptime_t?

rlubos reviewed (commented):
`net_timestamp_t` sounds ok.

fgrandel commented on include/zephyr/net/net_ktime.h:
```diff
@@ -0,0 +1,108 @@
+ * type, see the ongoing discussion in #60400 for details.
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_KTIME_H_
```
Fixed.

fgrandel commented on include/zephyr/net/ieee802154_radio.h:
```diff
@@ -382,77 +466,251 @@ struct ieee802154_config {
+	 * @param set true to set the filter, false to remove it
+	 * @param type the type of entity to be added/removed from the filter
+	 * list (a PAN ID or a source/destination address)
+	 * @param type the entity to be added/removed from the filter
```
Fixed.

fgrandel commented:
> `net_timestamp_t` sounds ok.

Renamed to net_time_t as you know - so that it also intuitively represent time spans. The result now feels much more natural to me, thanks for you feedback there.

jukkar reviewed (commented):
I cannot do now full review, but I faintly recall that the timestamp and txtime fields in net_pkt were suppose to be separate so that both can be enabled / used at the same time. Did you check this when doing the pr?

fgrandel commented:
> I cannot do now full review, but I faintly recall that the timestamp and txtime fields in net_pkt were suppose to be separate so that both can be enabled / used at the same time. Did you check this when doing the pr?

@jukkar Yes I did and I didn't find any use case where these would be used together or even where it theoretically makes sense to use them together. It also has to be mentioned that Linux has only one field as well if I'm not mistaken.

But of course I could have overlooked something. So it would be great if this result could be confirmed by others.

fgrandel commented:
@jukkar Oh and btw - it's not a problem if this PR is not merged quickly. I prefer thorough review over fast merge. :-) So if it's just a matter of doing review later, that is of course totally fine. You review would be very much appreciated.

jukkar reviewed (commented):
Ok, that is good news. I am currently having a vacation and travelling but I am trying to review the pr asap.

fgrandel commented:
@rlubos @edmont @jukkar The recently merged get_attr method caused merge conflicts which are now resolved. I also added more documentation to that new attribute feature, especially focusing on potential abuse. @tbursztyka This hopefully addresses the concerns that you had against introducing a getter. Please note that this is NOT a configuration getter but reserved to scalar/structured attributes that can neither be modeled as boolean HW CAPs nor as L2-driven configuration but originate from the driver itself (e.g. some of the PHY PIB attributes/constants).

jukkar reviewed (dismissed):
LGTM. Very nice to see extensive documentation of various structs. I think we need to add something to release notes as some api changes are there.

Commit:
doc: release-notes: document net subsys time consolidation

Documents the changes to the internal API in the release notes.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

fgrandel commented:
@jukkar Thanks for your review! Release notes added.

@jukkar @rlubos Think this is now ready for final approval and merge.

cfriedt reviewed (approved):
🪁
